### PR TITLE
#27 Added support for dropping database objects

### LINKED ISSUE #27: Add drop method to DatabaseObject
Add a `drop()` method to `DatabaseObject`

The method should drop the database object.

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #27

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -139,6 +139,98 @@ Covers:
...
 Needs: impl, itest
 
+## Dropping database objects
+
+### Dropping Generic Database Objects
+
+#### Dropping Users
+`dsn~dropping-users~1`
+
+Users can drop users using TDDB.
```
Comment (AnastasiiaSergienko) [inline doc/design.md:149]: Sounds like a war : D

Comment (AnastasiiaSergienko) [inline doc/design.md:149]:
```suggestion
Users can drop database users using TDDB.
```

File: doc/design.md [hunk 2]
```diff
@@ -139,6 +139,98 @@ Covers:
...
+Covers:
+
+* `req~dropping-exasol-database-objects~1`
+
+Needs: impl, itest
+
+### Dropping MySQL Database Objects
+`dsn~dropping-mysql-database-objects~1`
+
+MySQL has currently no database specific objects that could be dropped.
```
Comment (AnastasiiaSergienko) [inline doc/design.md:228]: Maybe you can remove this and the requirement if there are no real requirements. We can add it later if something comes up

File: pom.xml [hunk 3]
```diff
@@ -271,7 +271,7 @@
             <plugin>
                 <groupId>org.itsallcode</groupId>
                 <artifactId>openfasttrace-maven-plugin</artifactId>
-                <version>0.1.0</version>
+                <version>1.0.0</version>
```
Comment (AnastasiiaSergienko) [inline pom.xml:274]: Why did you change a major version? This branch looks as 0.2.0 to me

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java [hunk 4]
```diff
@@ -122,6 +149,15 @@ public void write(final VirtualSchema virtualSchema) {
         writeToObject(virtualSchema, builder.toString());
     }
 
+    /**
+     * Drop a Virtual Schema.
+     *
+     * @param virtualSchema Virtual Schema to drop.
+     */
+    public void drop(final VirtualSchema virtualSchema) {
+        writeToObject(virtualSchema, "DROP VIRTUAL SCHEMA " + virtualSchema.getFullyQualifiedName() + " CASCADE");
```
Comment (AnastasiiaSergienko) [inline src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java:158]: Why did you implement cascade only for virtual schemas? I think other Exasol drop commands should also be able to use cascade and it should be also decided by the user which one he prefers to use. And now it's hidden in the implementation

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]

File: doc/design.md (hunk 2 above)
Comment (jakobbraun) [reply doc/design.md:228]: I added it to creep track of the different dialects. So for all different dialect there is at least one design for the implementation of dropping.

File: pom.xml (hunk 3 above)
Comment (jakobbraun) [reply pom.xml:274]: It's just the version of the openfasttrace-maven-plugin. I did not change ther version number yet.

File: src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java (hunk 4 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java:158]: As far as i found out, DROP VIRTUAL SCHEMA always requires the CASCADE. Therefore I added it.

Comment (AnastasiiaSergienko) [reply src/main/java/com/exasol/dbbuilder/dialects/exasol/ExasolImmediateDatabaseObjectWriter.java:158]: Are you 100% sure? I also always use CASCADE for virtual schemas, but according to the documentation on the portal, it's an optional value and an example also doesn't have it: https://docs.exasol.com/sql/drop_schema.htm

Please, check again. If it's possible to use drop without cascade with virtual schema, I suggest having a usual drop and drop with the cascade (same for other structures)

**Review by AnastasiiaSergienko — CHANGES_REQUESTED** [review]
Don't forget to create changes file and add your changes there

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractImmediateDatabaseObjectWriter.java [hunk 5]
```diff
@@ -112,4 +112,17 @@ public void executeSqlFile(final Path... sqlFiles) {
             }
         }
     }
+
+    @Override
+    // [impl->dsn~dropping-tables~1]
+    public void drop(final Table table) {
+        final String sql = "DROP TABLE " + table.getFullyQualifiedName();
+        writeToObject(table, sql);
```
Comment (morazow) [inline src/main/java/com/exasol/dbbuilder/dialects/AbstractImmediateDatabaseObjectWriter.java:120]:
```suggestion
        writeToObject(table, "DROP TABLE " + table.getFullyQualifiedName());
```

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractSchema.java [hunk 6]
```diff
@@ -1,6 +1,7 @@
 package com.exasol.dbbuilder.dialects;
 
 import java.util.ArrayList;
+import java.util.Collections;
```
Comment (morazow) [inline src/main/java/com/exasol/dbbuilder/dialects/AbstractSchema.java:4]: Is this needed?

File: src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java [hunk 7]
```diff
@@ -72,4 +72,10 @@ public User grant(final GlobalPrivilege... privileges) {
         getWriter().write(this, privileges);
         return this;
     }
+
+    // [impl->dsn~dropping-users~1]
+    @Override
```
Comment (morazow) [inline src/main/java/com/exasol/dbbuilder/dialects/AbstractUser.java:77]:
```suggestion
    @Override
     // [impl->dsn~dropping-users~1]
```

File: src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java [hunk 8]
```diff
@@ -298,4 +321,26 @@ void testInsertIntoTable() {
                     exception);
         }
     }
+
+    @Override
+    protected Matcher<DatabaseObject> existsInDatabase() {
+        return new ExistsInDatabaseMatcher();
+    }
+
+    private class ExistsInDatabaseMatcher extends AbstractDatabaseObjectCreationAndDeletionIT.ExistsInDatabaseMatcher {
```
Comment (morazow) [inline src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:330]: Could make this static class?

File: src/test/java/com/exasol/dbbuilder/dialects/mysql/MySQLDatabaseObjectCreationAndDeletionIT.java [hunk 9]
```diff
@@ -90,11 +84,34 @@ void testGrantSchemaPrivilegeToUser() throws InterruptedException {
                 () -> assertUserHasSchemaPrivilege(user.getName(), schema.getName(), "Delete_priv"));
     }
 
-    private void assertUserHasSchemaPrivilege(final String username, final String objectName, final String columnName)
-            throws AssertionError {
-        final String sql = "SELECT `" + columnName + "` FROM mysql.db WHERE User = '" + username + "' AND Db = '"
-                + objectName + "'";
-        assertObjectExistsInDatabase(sql, "Y");
+    private class ExistsInDatabaseMatcher extends AbstractDatabaseObjectCreationAndDeletionIT.ExistsInDatabaseMatcher {
```
Comment (morazow) [inline src/test/java/com/exasol/dbbuilder/dialects/mysql/MySQLDatabaseObjectCreationAndDeletionIT.java:87]: Same here, make static and provide admin connection as parameter.

**Review by morazow — CHANGES_REQUESTED** [review]

File: src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java (hunk 8 above)
Comment (jakobbraun) [reply src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:330]: I thought it's nicer that way since existsInDatabase() is just like a sentence. The performance loss is acceptable in test code... So what is the problem with a non-static class?

Comment (morazow) [reply src/test/java/com/exasol/dbbuilder/dialects/exasol/ExasolDatabaseObjectCreationAndDeletionIT.java:328]: It is one of the best practices. The main reasons:
> Without static member classes, each instance will have extra reference to its enclosing instance. This reference costs time and space, and result enclosing instance to be retained when it is eligible for garbage collection.
You can also read Item 22 in Effective Java.

Yes, in this test code maybe it is okay, though.

I think you can keep the method `existsInDatabase()` as is, and inside provide parameter to the new instance of the inner class.

**Review by morazow — APPROVED** [review]
